File: Advent-of-Code-2021/day04.ts
import * as utils from "./utils.ts";

const lines = utils.readDay(4);
const numbers: Readonly<string[]> = lines.shift()!.split(",");

const tables: utils.DeepReadonly<string[][][]> = utils.chunks(
  lines.filter((line) => line).map((line) => line.trim().split(/\s+/)),
  5
);
const uncalledNumbers: string[][][] = structuredClone(tables);

const rowsCounts = tables.map(() => utils.range(5).map(() => 0));
const columnCounts: number[][] = structuredClone(rowsCounts);

const numbersIndex: { [K in string]: [number, number, number][] } = {};
tables.forEach((table, tableIndex) =>
  table.forEach((row, rowIndex) =>
    row.forEach((value, colIndex) => {
      if (numbersIndex[value] === undefined) numbersIndex[value] = [];
      numbersIndex[value].push([tableIndex, rowIndex, colIndex]);
    })
  )
);
let winningNumber: string;
let winningUncalledNumbers: number[];
let losingNumber: string;
let losingUncalledNumbers: number[];
const wonTables: Set<number> = new Set();

numbers.forEach((number) => {
  numbersIndex[number].forEach(([table, row, col]) => {
    if (wonTables.has(table)) {
      return;
    }
    rowsCounts[table][row] += 1;
    columnCounts[table][col] += 1;
    uncalledNumbers[table][row][col] = "";
    if (rowsCounts[table][row] == 5 || columnCounts[table][col] == 5) {
      if (winningNumber === undefined) {
        winningNumber = number;
        winningUncalledNumbers = uncalledNumbers[table].flat().map(Number);
      }
      losingNumber = number;
      losingUncalledNumbers = uncalledNumbers[table].flat().map(Number);
      wonTables.add(table);
    }
  });
});
console.log(utils.sum(winningUncalledNumbers!) * Number(winningNumber!));
console.log(utils.sum(losingUncalledNumbers!) * Number(losingNumber!));

File: Advent-of-Code-2021/day04.py
import itertools
from collections import defaultdict

lines = open("day04.txt", "r").read().splitlines(keepends=False)
numbers = lines.pop(0).split(",")
lines.pop(0)
lines.append("")
tables: list[list[list[str]]] = []
current_table: list[list[str]] = []
for line in lines:
    if not line:
        tables.append(current_table)
        current_table = []
    else:
        current_table.append(line.split())

numbers_index: defaultdict[str, list[tuple[int, int, int]]] = defaultdict(list)
for table_i, table in enumerate(tables):
    for row, col in itertools.product(range(5), range(5)):
        numbers_index[table[row][col]].append((table_i, row, col))

winningNumber = None
losingNumber = None
uncalled_numbers_sum = None

won_tables: set[int] = set()
for number in numbers:
    for (table_i, row, col) in numbers_index[number]:
        if table_i in won_tables:
            continue
        table = tables[table_i]
        table[row][col] = ""
        if not (
            all(table[row][i] == "" for i in range(5))
            or all(table[i][col] == "" for i in range(5))
        ):
            continue
        won_tables.add(table_i)
        uncalled_numbers_sum = sum(int(n) for row in table for n in row if n)
        if winningNumber is None:
            winningNumber = int(number)
            print(winningNumber * uncalled_numbers_sum)
        losingNumber = int(number)

assert losingNumber is not None and uncalled_numbers_sum is not None
print(losingNumber * uncalled_numbers_sum)
